src/query/query_visitor.rs

```diff
@@ -0,0 +1,227 @@
+        },
+    }
+}
+
+fn walk_fragment_definition<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast FragmentDefinition) {
+    visitor.visit_fragment_definition(node);
+    walk_selection_set(visitor, &node.selection_set);
+}
+
+fn walk_operation_definition<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast OperationDefinition) {
+    use super::ast::OperationDefinition::*;
+
+    visitor.visit_operation_definition(node);
+    match node {
+        SelectionSet(inner) => {
+            visitor.visit_selection_set(inner);
+            walk_selection_set(visitor, inner);
+        }
+        Query(inner) => {
+            visitor.visit_query(inner);
+            walk_query(visitor, inner);
+        }
+        Mutation(inner) => {
+            visitor.visit_mutation(inner);
+            walk_mutation(visitor, inner);
+        }
+        Subscription(inner) => {
+            visitor.visit_subscription(inner);
+            walk_subscription(visitor, inner);
+        }
+    }
+}
+
+fn walk_query<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Query) {
+    visitor.visit_query(node);
+
+    for var_def in &node.variable_definitions {
+        visitor.visit_variable_definition(var_def);
+        walk_variable_definition(visitor, var_def);
+    }
+
+    visitor.visit_selection_set(&node.selection_set);
+    walk_selection_set(visitor, &node.selection_set);
+}
+
+fn walk_mutation<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Mutation) {
+    visitor.visit_mutation(node);
+
+    for var_def in &node.variable_definitions {
+        visitor.visit_variable_definition(var_def);
+        walk_variable_definition(visitor, var_def);
+    }
+
+    visitor.visit_selection_set(&node.selection_set);
+    walk_selection_set(visitor, &node.selection_set);
+}
+
+fn walk_subscription<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Subscription) {
+    visitor.visit_subscription(node);
+
+    for var_def in &node.variable_definitions {
+        visitor.visit_variable_definition(var_def);
+        walk_variable_definition(visitor, var_def);
+    }
+
+    visitor.visit_selection_set(&node.selection_set);
+    walk_selection_set(visitor, &node.selection_set);
+}
+
+fn walk_selection_set<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast SelectionSet) {
+    visitor.visit_selection_set(node);
+
+    for selection in &node.items {
+        visitor.visit_selection(selection);
+        walk_selection(visitor, selection);
+    }
+}
+
+fn walk_variable_definition<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast VariableDefinition) {
+    visitor.visit_variable_definition(node)
+}
+
+fn walk_selection<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Selection) {
+    use super::ast::Selection::*;
+
+    visitor.visit_selection(node);
+    match node {
+        Field(inner) => {
+            visitor.visit_field(inner);
+            walk_field(visitor, inner);
```

Comment:
I don't understand this pattern. You first do `visit_field` and then `walk_field` which calls `visit_field` again. Same with fragments. Am I misunderstanding something?

Reply:
I have a similar question. It seems that every walk function calls visit_node and then then calls a walk_node function that calls the same visit_node. For example `walk_subscription` calls `visitor.visit_selection_set` and then cals `walk_selection_set` which in turn calls `visitor.visit_selection_set` a second time. 

Is that how it's supposed to work? I thought you only need to visit once.